### PR TITLE
HDFS-15498. Show snapshots deletion status in snapList cmd.

### PR DESCRIPTION
**Pull request by bshashikant**
please see https://issues.apache.org/jira/browse/HDFS-15498

### CODE REVIEW

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |  29m 41s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +0 :ok: |  prototool  |   0m  1s |  prototool was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 1 new or modified test files.  |
||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |   1m  4s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  21m 45s |  trunk passed  |
| +1 :green_heart: |  compile  |   4m 13s |  trunk passed with JDK Ubuntu-11.0.8+10-post-Ubuntu-0ubuntu118.04.1  |
| +1 :green_heart: |  compile  |   3m 49s |  trunk passed with JDK Private Build-1.8.0_252-8u252-b09-1~18.04-b09  |
| +1 :green_heart: |  checkstyle  |   0m 59s |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   2m  5s |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  18m 45s |  branch has no errors when building and testing our client artifacts.  |
| -1 :x: |  javadoc  |   0m 35s |  hadoop-hdfs-client in trunk failed with JDK Ubuntu-11.0.8+10-post-Ubuntu-0ubuntu118.04.1.  |
| -1 :x: |  javadoc  |   0m 32s |  hadoop-hdfs in trunk failed with JDK Ubuntu-11.0.8+10-post-Ubuntu-0ubuntu118.04.1.  |
| +1 :green_heart: |  javadoc  |   1m 10s |  trunk passed with JDK Private Build-1.8.0_252-8u252-b09-1~18.04-b09  |
| +0 :ok: |  spotbugs  |   3m  8s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   5m 33s |  trunk passed  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 22s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   1m 55s |  the patch passed  |
| +1 :green_heart: |  compile  |   4m  8s |  the patch passed with JDK Ubuntu-11.0.8+10-post-Ubuntu-0ubuntu118.04.1  |
| +1 :green_heart: |  cc  |   4m  8s |  the patch passed  |
| +1 :green_heart: |  javac  |   4m  8s |  the patch passed  |
| +1 :green_heart: |  compile  |   3m 44s |  the patch passed with JDK Private Build-1.8.0_252-8u252-b09-1~18.04-b09  |
| +1 :green_heart: |  cc  |   3m 44s |  the patch passed  |
| +1 :green_heart: |  javac  |   3m 44s |  the patch passed  |
| -0 :warning: |  checkstyle  |   0m 55s |  hadoop-hdfs-project: The patch generated 1 new + 60 unchanged - 0 fixed = 61 total (was 60)  |
| +1 :green_heart: |  mvnsite  |   1m 59s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedclient  |  15m 28s |  patch has no errors when building and testing our client artifacts.  |
| -1 :x: |  javadoc  |   0m 33s |  hadoop-hdfs-client in the patch failed with JDK Ubuntu-11.0.8+10-post-Ubuntu-0ubuntu118.04.1.  |
| -1 :x: |  javadoc  |   0m 29s |  hadoop-hdfs in the patch failed with JDK Ubuntu-11.0.8+10-post-Ubuntu-0ubuntu118.04.1.  |
| +1 :green_heart: |  javadoc  |   1m  5s |  the patch passed with JDK Private Build-1.8.0_252-8u252-b09-1~18.04-b09  |
| -1 :x: |  findbugs  |   2m 34s |  hadoop-hdfs-project/hadoop-hdfs-client generated 1 new + 0 unchanged - 0 fixed = 1 total (was 0)  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   1m 55s |  hadoop-hdfs-client in the patch passed.  |
| -1 :x: |  unit  | 109m  7s |  hadoop-hdfs in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 36s |  The patch does not generate ASF License warnings.  |
|  |   | 239m  0s |   |


| Reason | Tests |
|-------:|:------|
| FindBugs | module:hadoop-hdfs-project/hadoop-hdfs-client |
|  |  There is an apparent infinite recursive loop in org.apache.hadoop.hdfs.protocol.SnapshotStatus$Bean.isDELETED()  At SnapshotStatus.java:recursive loop in org.apache.hadoop.hdfs.protocol.SnapshotStatus$Bean.isDELETED()  At SnapshotStatus.java:[line 220] |
| Failed junit tests | hadoop.hdfs.server.namenode.TestNameNodeRetryCacheMetrics |
|   | hadoop.fs.contract.hdfs.TestHDFSContractMultipartUploader |
|   | hadoop.hdfs.server.blockmanagement.TestBlockTokenWithDFSStriped |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.40 ServerAPI=1.40 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-2181/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/2181 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle cc prototool |
| uname | Linux 2bcb330357a9 4.15.0-101-generic #102-Ubuntu SMP Mon May 11 10:07:26 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / 05b3337a460 |
| Default Java | Private Build-1.8.0_252-8u252-b09-1~18.04-b09 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.8+10-post-Ubuntu-0ubuntu118.04.1 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_252-8u252-b09-1~18.04-b09 |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-2181/1/artifact/out/branch-javadoc-hadoop-hdfs-project_hadoop-hdfs-client-jdkUbuntu-11.0.8+10-post-Ubuntu-0ubuntu118.04.1.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-2181/1/artifact/out/branch-javadoc-hadoop-hdfs-project_hadoop-hdfs-jdkUbuntu-11.0.8+10-post-Ubuntu-0ubuntu118.04.1.txt |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-2181/1/artifact/out/diff-checkstyle-hadoop-hdfs-project.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-2181/1/artifact/out/patch-javadoc-hadoop-hdfs-project_hadoop-hdfs-client-jdkUbuntu-11.0.8+10-post-Ubuntu-0ubuntu118.04.1.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-2181/1/artifact/out/patch-javadoc-hadoop-hdfs-project_hadoop-hdfs-jdkUbuntu-11.0.8+10-post-Ubuntu-0ubuntu118.04.1.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-2181/1/artifact/out/new-findbugs-hadoop-hdfs-project_hadoop-hdfs-client.html |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-2181/1/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-2181/1/testReport/ |
| Max. process+thread count | 2785 (vs. ulimit of 5500) |
| modules | C: hadoop-hdfs-project/hadoop-hdfs-client hadoop-hdfs-project/hadoop-hdfs U: hadoop-hdfs-project |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-2181/1/console |
| versions | git=2.17.1 maven=3.6.0 findbugs=4.0.6 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

File: hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/protocol/SnapshotStatus.java [hunk 1]
```diff
@@ -207,6 +215,10 @@ public String getOwner() {
     public String getGroup() {
       return group;
     }
+
+    public boolean isDELETED () {
```
[the remark this answers is not in the document]
Comment (mukul1987) [reply hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/protocol/SnapshotStatus.java:219]: This function can be removed, as this has no callers and is causing recuriosn.

File: hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/namenode/snapshot/TestListSnapshot.java [hunk 2]
```diff
@@ -128,7 +132,15 @@ public void testListSnapshot() throws Exception {
...
-    // There are now 2 snapshots for dir1
+    // There are now 2 active snapshots for dir1 and one is marked deleted
+    assertEquals(3, snapshotStatuses.length);
+    assertTrue(snapshotStatuses[2].isDeleted());
+    assertFalse(snapshotStatuses[1].isDeleted());
+    assertFalse(snapshotStatuses[0].isDeleted());
+    // delete the 1st snapshot
+    hdfs.deleteSnapshot(dir1, "s0");
+    snapshotStatuses = hdfs.getSnapshotListing(dir1);
+    // There are now 2 snapshots now as the 1st one is deleted in order
```
Comment (mukul1987) [inline hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/namenode/snapshot/TestListSnapshot.java:143]: Let's add assert that S0 is marked as deleted here.

Comment (bshashikant) [reply hadoop-hdfs-project/hadoop-hdfs/src/test/java/org/apache/hadoop/hdfs/server/namenode/snapshot/TestListSnapshot.java:143]: S0 is the first snapshot in the list. It won't be marked deleted.

File: hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/protocol/SnapshotStatus.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (bshashikant) [reply hadoop-hdfs-project/hadoop-hdfs-client/src/main/java/org/apache/hadoop/hdfs/protocol/SnapshotStatus.java]: Addressed

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| +0 :ok: |  reexec  |   1m 50s |  Docker mode activated.  |
||| _ Prechecks _ |
| +1 :green_heart: |  dupname  |   0m  0s |  No case conflicting files found.  |
| +0 :ok: |  prototool  |   0m  1s |  prototool was not available.  |
| +1 :green_heart: |  @author  |   0m  0s |  The patch does not contain any @author tags.  |
| +1 :green_heart: |  test4tests  |   0m  0s |  The patch appears to include 1 new or modified test files.  |
||| _ trunk Compile Tests _ |
| +0 :ok: |  mvndep  |   1m  2s |  Maven dependency ordering for branch  |
| +1 :green_heart: |  mvninstall  |  22m 20s |  trunk passed  |
| +1 :green_heart: |  compile  |   4m 20s |  trunk passed with JDK Ubuntu-11.0.8+10-post-Ubuntu-0ubuntu118.04.1  |
| +1 :green_heart: |  compile  |   3m 53s |  trunk passed with JDK Private Build-1.8.0_252-8u252-b09-1~18.04-b09  |
| +1 :green_heart: |  checkstyle  |   0m 58s |  trunk passed  |
| +1 :green_heart: |  mvnsite  |   2m  4s |  trunk passed  |
| +1 :green_heart: |  shadedclient  |  18m 35s |  branch has no errors when building and testing our client artifacts.  |
| -1 :x: |  javadoc  |   0m 38s |  hadoop-hdfs-client in trunk failed with JDK Ubuntu-11.0.8+10-post-Ubuntu-0ubuntu118.04.1.  |
| -1 :x: |  javadoc  |   0m 31s |  hadoop-hdfs in trunk failed with JDK Ubuntu-11.0.8+10-post-Ubuntu-0ubuntu118.04.1.  |
| +1 :green_heart: |  javadoc  |   1m 11s |  trunk passed with JDK Private Build-1.8.0_252-8u252-b09-1~18.04-b09  |
| +0 :ok: |  spotbugs  |   3m  6s |  Used deprecated FindBugs config; considering switching to SpotBugs.  |
| +1 :green_heart: |  findbugs  |   5m 31s |  trunk passed  |
||| _ Patch Compile Tests _ |
| +0 :ok: |  mvndep  |   0m 21s |  Maven dependency ordering for patch  |
| +1 :green_heart: |  mvninstall  |   1m 55s |  the patch passed  |
| +1 :green_heart: |  compile  |   4m  9s |  the patch passed with JDK Ubuntu-11.0.8+10-post-Ubuntu-0ubuntu118.04.1  |
| +1 :green_heart: |  cc  |   4m  9s |  the patch passed  |
| +1 :green_heart: |  javac  |   4m  9s |  the patch passed  |
| +1 :green_heart: |  compile  |   3m 45s |  the patch passed with JDK Private Build-1.8.0_252-8u252-b09-1~18.04-b09  |
| +1 :green_heart: |  cc  |   3m 45s |  the patch passed  |
| +1 :green_heart: |  javac  |   3m 45s |  the patch passed  |
| +1 :green_heart: |  checkstyle  |   0m 53s |  the patch passed  |
| +1 :green_heart: |  mvnsite  |   1m 56s |  the patch passed  |
| +1 :green_heart: |  whitespace  |   0m  0s |  The patch has no whitespace issues.  |
| +1 :green_heart: |  shadedclient  |  15m 43s |  patch has no errors when building and testing our client artifacts.  |
| -1 :x: |  javadoc  |   0m 34s |  hadoop-hdfs-client in the patch failed with JDK Ubuntu-11.0.8+10-post-Ubuntu-0ubuntu118.04.1.  |
| -1 :x: |  javadoc  |   0m 29s |  hadoop-hdfs in the patch failed with JDK Ubuntu-11.0.8+10-post-Ubuntu-0ubuntu118.04.1.  |
| +1 :green_heart: |  javadoc  |   1m  7s |  the patch passed with JDK Private Build-1.8.0_252-8u252-b09-1~18.04-b09  |
| +1 :green_heart: |  findbugs  |   5m 42s |  the patch passed  |
||| _ Other Tests _ |
| +1 :green_heart: |  unit  |   1m 58s |  hadoop-hdfs-client in the patch passed.  |
| -1 :x: |  unit  | 109m 44s |  hadoop-hdfs in the patch passed.  |
| +1 :green_heart: |  asflicense  |   0m 36s |  The patch does not generate ASF License warnings.  |
|  |   | 212m 44s |   |


| Reason | Tests |
|-------:|:------|
| Failed junit tests | hadoop.hdfs.server.sps.TestExternalStoragePolicySatisfier |
|   | hadoop.hdfs.TestGetFileChecksum |
|   | hadoop.hdfs.server.namenode.TestNameNodeRetryCacheMetrics |
|   | hadoop.fs.contract.hdfs.TestHDFSContractMultipartUploader |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | ClientAPI=1.40 ServerAPI=1.40 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-2181/2/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/2181 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle cc prototool |
| uname | Linux d245145b5086 4.15.0-91-generic #92-Ubuntu SMP Fri Feb 28 11:09:48 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / e756fe35909 |
| Default Java | Private Build-1.8.0_252-8u252-b09-1~18.04-b09 |
| Multi-JDK versions | /usr/lib/jvm/java-11-openjdk-amd64:Ubuntu-11.0.8+10-post-Ubuntu-0ubuntu118.04.1 /usr/lib/jvm/java-8-openjdk-amd64:Private Build-1.8.0_252-8u252-b09-1~18.04-b09 |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-2181/2/artifact/out/branch-javadoc-hadoop-hdfs-project_hadoop-hdfs-client-jdkUbuntu-11.0.8+10-post-Ubuntu-0ubuntu118.04.1.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-2181/2/artifact/out/branch-javadoc-hadoop-hdfs-project_hadoop-hdfs-jdkUbuntu-11.0.8+10-post-Ubuntu-0ubuntu118.04.1.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-2181/2/artifact/out/patch-javadoc-hadoop-hdfs-project_hadoop-hdfs-client-jdkUbuntu-11.0.8+10-post-Ubuntu-0ubuntu118.04.1.txt |
| javadoc | https://builds.apache.org/job/hadoop-multibranch/job/PR-2181/2/artifact/out/patch-javadoc-hadoop-hdfs-project_hadoop-hdfs-jdkUbuntu-11.0.8+10-post-Ubuntu-0ubuntu118.04.1.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-2181/2/artifact/out/patch-unit-hadoop-hdfs-project_hadoop-hdfs.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-2181/2/testReport/ |
| Max. process+thread count | 3116 (vs. ulimit of 5500) |
| modules | C: hadoop-hdfs-project/hadoop-hdfs-client hadoop-hdfs-project/hadoop-hdfs U: hadoop-hdfs-project |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-2181/2/console |
| versions | git=2.17.1 maven=3.6.0 findbugs=4.0.6 |
| Powered by | Apache Yetus 0.12.0 https://yetus.apache.org |


This message was automatically generated.

**Comment by hemanthboyina** [comment]
changes looks good @bshashikant  , can we extend snapshot deletion status in "ls" command ?
as -ls /user/.snapshot details will also give all snapshot related details for /user

**Comment by mukul1987** [comment]
@hemanthboyina, we are not modifying the "ls" command because the changes will be backward incompatible.

**Comment by mukul1987** [comment]
The patch looks good to me. I will commit them shortly.

**Pull request merged**